cpp/src/replace/replace.cu

```diff
@@ -13,31 +16,528 @@
+                           const T* __restrict__       d_replacement_values,
+                           cudf::bitmask_type const * __restrict__ replacement_valid)
+   {
+     auto found_ptr = thrust::find(thrust::seq, values_to_replace_begin,
+                                      values_to_replace_end, input_data[idx]);
+     T new_value{0};
+     bool output_is_valid{true};
+
+     if (found_ptr != values_to_replace_end) {
+       auto d = thrust::distance(values_to_replace_begin, found_ptr);
+       new_value = d_replacement_values[d];
+       if (replacement_has_nulls) {
+         output_is_valid = cudf::bit_is_set(replacement_valid, d);
+       }
+     } else {
+       new_value = input_data[idx];
+     }
+     return thrust::make_pair(new_value, output_is_valid);
+   }
+
+  /* --------------------------------------------------------------------------*/
+  /**
+   * @brief Kernel that replaces elements from `output_data` given the following
+   *        rule: replace all `values_to_replace[i]` in [values_to_replace_begin`,
+   *        `values_to_replace_end`) present in `output_data` with `d_replacement_values[i]`.
+   *
+   * @tparam input_has_nulls `true` if output column has valid mask, `false` otherwise
+   * @tparam replacement_has_nulls `true` if replacement_values column has valid mask, `false` otherwise
+   * The input_has_nulls and replacement_has_nulls template parameters allows us to specialize
+   * this kernel for the different scenario for performance without writing different kernel.
+   *
+   * @param[in] input_data Device array with the data to be modified
+   * @param[in] input_valid Valid mask associated with input_data
+   * @param[out] output_data Device array to store the data from input_data
+   * @param[out] output_valid Valid mask associated with output_data
+   * @param[out] output_valid_count #valid in output column
+   * @param[in] nrows # rows in `output_data`
+   * @param[in] values_to_replace_begin Device pointer to the beginning of the sequence
+   * of old values to be replaced
+   * @param[in] values_to_replace_end  Device pointer to the end of the sequence
+   * of old values to be replaced
+   * @param[in] d_replacement_values Device array with the new values
+   * @param[in] replacement_valid Valid mask associated with d_replacement_values
+   *
+   * @returns
+   */
+  /* ----------------------------------------------------------------------------*/
+  template <class T,
+            bool input_has_nulls, bool replacement_has_nulls>
+  __global__
+  void replace_kernel(cudf::column_device_view input,
+                      cudf::mutable_column_device_view output,
+                      cudf::size_type * __restrict__    output_valid_count,
+                      cudf::size_type                   nrows,
+                      cudf::column_device_view values_to_replace,
+                      cudf::column_device_view replacement)
+  {
+  const T* __restrict__ input_data = input.data<T>();
```

Comment:
Pulling out all these pointers really defeats the purpose of providing `column_device_view`s with device member functions.

Reply:
What change would you suggest here?

Reply:
Don't pull out the raw pointers and use the object's methods directly like `element, is_null, is_valid`, etc.

Reply:
Done.